app/grandchallenge/components/serializers.py

```diff
@@ -26,32 +30,130 @@ class SimpleImageSerializer(serializers.ModelSerializer):
+    interface = ComponentInterfaceSerializer(read_only=True)
+    upload_pk = serializers.CharField(write_only=True, required=False)
 
     class Meta:
         model = ComponentInterfaceValue
         fields = [
+            "interface_slug",
             "interface",
             "value",
             "file",
             "image",
             "pk",
+            "upload_pk",
         ]
 
+    def validate(self, attrs):  # noqa: C901
+        interface = ComponentInterface.objects.get(
+            slug=attrs["interface_slug"]
+        )
+        attrs.pop("interface_slug")
+        attrs["interface"] = interface
+
+        def get_value():
+            value = attrs.get("value", None)
+            if not value:
+                raise serializers.ValidationError(
+                    f"Value is required for interface kind {interface.kind}"
+                )
+            return value
+
+        def validate_simple():
+            kind_to_type = {
+                ComponentInterface.Kind.INTEGER: int,
+                ComponentInterface.Kind.STRING: str,
+                ComponentInterface.Kind.BOOL: bool,
+                ComponentInterface.Kind.FLOAT: float,
+            }
+            value = get_value()
+            if not isinstance(value, kind_to_type.get(interface.kind)):
+                raise serializers.ValidationError(
+                    f"Type of {value} does not match interface kind {interface.kind}"
+                )
+
+        def validate_annotations():
+            value = get_value()
+            allowed_types = [{"$ref": f"#/definitions/{interface.kind}"}]
+
+            JSONSchemaValidator(
+                schema={**ANSWER_TYPE_SCHEMA, "anyOf": allowed_types}
+            )(value)
+
+        def validate_image():
+            # either image or upload_pk should be provided
+            if not any(key in attrs for key in ("upload_pk", "image")):
+                raise serializers.ValidationError(
+                    f"Upload_pk or image are required for interface kind {interface.kind}"
+                )
+
+        if interface.kind in InterfaceKind.interface_type_simple():
+            validate_simple()
+        if interface.kind in InterfaceKind.interface_type_annotation():
+            validate_annotations()
+        if interface.kind in InterfaceKind.interface_type_image():
+            validate_image()
+
+        return attrs
+
+    def validate_upload_pk(self, value):
+        session = RawImageUploadSession.objects.get(pk=value)
+
+        user = self.context.get("user")
+
+        if not user.has_perm("view_rawimageuploadsession", session):
+            raise serializers.ValidationError(
+                f"User does not have permission to use raw image upload session {value}"
+            )
+
+        if session.status is not RawImageUploadSession.PENDING:
+            raise serializers.ValidationError(
+                f"Upload session {value} is not ready to be used"
+            )
```

Comment:
What would happen if the session has another status like failed, cancelled, requeued?

Reply:
There's a check on session status later in the process, so the job would fail later on.